src/go/util/httppattern/sort_impl.go

```diff
@@ -0,0 +1,261 @@
+		if val, ok := hn.ResultMap[httpMethod]; ok {
+			if markDuplicate {
+				val.isMultiple = true
+			}
+			return false
+		}
+		hn.ResultMap[httpMethod] = &lookupResult{
+			data:       methodData,
+			isMultiple: false,
+		}
+		return true
+	}
+
+	curSeg := pathParts[pathPartsIdxCur]
+
+	if _, ok := hn.Children[curSeg]; !ok {
+		hn.Children[curSeg] = newHttpPatternTrieNode()
+	}
+
+	child, _ := hn.Children[curSeg]
+	if curSeg == WildCardPathKey {
+		child.WildCard = true
+	}
+
+	return child.insertTemplate(pathParts, pathPartsIdxCur+1, httpMethod, methodData, markDuplicate)
+}
+
+func (hn *httpPatternTrieNode) insertPath(pathParts []string, httpMethod string, methodData *methodData, markDuplicate bool) bool {
+	return hn.insertTemplate(pathParts, 0, httpMethod, methodData, markDuplicate)
+}
+
+// Traverse the httpPatternTrie in matching order and add the visited method in result.
+func (hn *httpPatternTrieNode) traverse(result *MethodSlice) {
+	appendMethodOnCurrentNode := func() {
+
+		// Sort the method in alphabet order to generate deterministic sequence for better unit testing.
+		var sortedKeys []string
+		// Put the wildcard method in the end.
+		var wildMethodResult *lookupResult
+		for key, val := range hn.ResultMap {
+			if key == HttpMethodWildCard {
+				wildMethodResult = val
+				continue
+			}
+			sortedKeys = append(sortedKeys, key)
+		}
+		sort.Strings(sortedKeys)
+
+		for _, key := range sortedKeys {
+			if val, ok := hn.ResultMap[key]; ok {
+				result.appendMethod(val.data.Method)
+			}
+		}
+
+		// Put the wildcard method in the end.
+		if wildMethodResult != nil {
+			result.appendMethod(wildMethodResult.data.Method)
+		}
+
+	}
+
+	traverseChildren := func() {
+		var singleParameterChild *httpPatternTrieNode
+		var wildCardPathPartChild *httpPatternTrieNode
+		var wildCardPathChild *httpPatternTrieNode
+		var exactMatchChildKey []string
+		for key, child := range hn.Children {
+			switch key {
+			case SingleParameterKey:
+				singleParameterChild = child
+			case WildCardPathPartKey:
+				wildCardPathPartChild = child
+			case WildCardPathKey:
+				wildCardPathChild = child
+			default:
+				exactMatchChildKey = append(exactMatchChildKey, key)
+			}
+		}
+
+		// Visit exact match children first.
+		// Sort the child keys to generate deterministic sequence for better unit testing.
+		sort.Strings(exactMatchChildKey)
+		for _, key := range exactMatchChildKey {
+			if child, ok := hn.Children[key]; ok {
+				child.traverse(result)
+			}
+		}
+
+		// Visit vague match children after.
+		for _, child := range []*httpPatternTrieNode{singleParameterChild, wildCardPathPartChild, wildCardPathChild} {
+			if child != nil {
+				child.traverse(result)
+			}
+		}
+	}
+
+	// If the current node is wildcard(**), its children has higher priority.
+	if hn.WildCard {
+		// Pre-order traverse.
+		traverseChildren()
+		// Post-order traverse.
+		appendMethodOnCurrentNode()
+	} else {
+		appendMethodOnCurrentNode()
```

Comment:
can we always do traverseChidlren first for all cases?

Reply:
Not quite understand your examples.

If there is wildcard underneath, it will cause problem.

For example, it will make like:
/a/**
/a

apparently, it should be
/a
/a/**

where `appendMethodOnCurrentNode()` should be called first

Reply:
My understanding is:  **  mean 1 or more.  so

1) /a/**
2) /a

Above order should be.   "/a" should not match 1)

Reply:
Refer the [spec](https://cloud.google.com/endpoints/docs/grpc-service-config/reference/rpc/google.api#path-template-syntax)
`The syntax ** matches zero or more URL path segments, which must be the last part of the URL path except the Verb`

Reply:
Well,  in actual implementation (both regex and path_matcher),  it means "1 and more".  You can try it.

"/a" will not match pattern "/a/**",    and regex for it (it has second slash). 

Well,  I just want you to justify why we need two kind of sorting order, one for double_wildcard and another for others.

Reply:
OK, I take back the previous example. 

How about 
/**
/**/a (not allowed in spec but we support)
/**:verb(this is allowed)

We have to put /** in the end as it is more vague.

Reply:
Yeap,  only in /** case, it means 0 or more. Agreed that /** should be after /.
In c++, we have special code to handle root path "/".   Do we need that here?

OK,  could you add comment on this special case to justify that order has to be this way?
Could you add test cases for this special case?